File: src/main/java/us/dison/compactmachines/api/crafting/recipes/layers/CCRecipeBlocks.java
package us.dison.compactmachines.api.crafting.recipes.layers;

import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Box;
import net.minecraft.util.math.Vec3i;
import net.minecraft.block.BlockState;
import java.util.Optional; 
import java.util.stream.Stream;
import java.util.Map;
public interface CCRecipeBlocks {
	Optional<String> getComponentAtPosition(BlockPos relative);
	Optional<BlockState> getStateAtPosition(BlockPos relative);
	Stream<BlockPos> getPositions();
	int getNumberKnownComponents();
	void rebuildComponentTotals();
	Map<String, Integer> getKnownComponentTotals();
	Box getSourceBounds();
	boolean allIdentified();
	Stream<BlockPos> getUnmappedPositions();
	Stream<BlockPos> getPositionsForComponent(String component);
	Box getFilledBounds();
	CCRecipeBlocks slice(Box bounds);
	CCRecipeBlocks offset(Vec3i amount);
	default CCRecipeBlocks below(int offset) {
		return offset(new Vec3i(0, -offset, 0));
	} 
	default CCRecipeBlocks above(int offset) {
		return offset(new Vec3i(0, offset, 0));
	}
	default CCRecipeBlocks normalize() {
		Box sb = getSourceBounds();
		BlockPos offset = new BlockPos(-sb.minX, -sb.minY, -sb.minZ);
		return offset(offset);
	}

}

File: src/main/scala/us/dison/compactmachines/crafting/recipes/layers/givens/ccRecipeBlocks.scala
package us.dison.compactmachines.crafting.recipes.layers.givens

import us.dison.compactmachines.*
import net.minecraft.util.math.{BlockBox, Box}
import us.dison.compactmachines.api.crafting.recipes.layers.IRecipeBlocks

import java.util.stream.Collectors

object ccRecipeBlocks {
extension (c: IRecipeBlocks) {
  def componentAtPosition(relative: BlockPos): Option[String] =
    c.getComponentAtPosition(relative).toScala
  def stateAtPosition(relative: BlockPos): Option[BlockState] =
    c.getStateAtPosition(relative).toScala
  def positions: List[BlockPos] =
    c.getPositions.collect(Collectors.toList).asScala.toList
  def numberKnownComponents: Int =
    c.getNumberKnownComponents
  def knownComponentTotals: Map[String, Int] =
    c.getKnownComponentTotals.asScala.toMap.view.mapValues(_.toInt).toMap
  def sourceBounds: BlockBox =
    c.getSourceBounds
  def unmappedPositions: List[BlockPos] =
    c.getUnmappedPositions.toScala(List)
  def positionsForComponent(component: String): List[BlockPos] =
    c.getPositionsForComponent(component).toScala(List)
  def filledBounds: BlockBox = c.getFilledBounds
}
}

